k8s/operators/src/pool/diskpool/crd/v1beta3.rs

```diff
@@ -0,0 +1,246 @@
+use kube::CustomResource;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+#[cfg(feature = "openapi")]
+use openapi::models::{pool_status::PoolStatus as RestPoolStatus, Pool};
+
+#[derive(
+    CustomResource, Serialize, Deserialize, Default, Debug, Eq, PartialEq, Clone, JsonSchema,
+)]
+#[kube(
+    group = "openebs.io",
+    version = "v1beta3",
+    kind = "DiskPool",
+    plural = "diskpools",
+    // The name of the struct that gets created that represents a resource
+    namespaced,
+    status = "DiskPoolStatus",
+    derive = "PartialEq",
+    derive = "Default",
+    shortname = "dsp",
+    printcolumn = r#"{ "name":"node", "type":"string", "description":"node the pool is on", "jsonPath":".spec.node"}"#,
+    printcolumn = r#"{ "name":"state", "type":"string", "description":"dsp cr state", "jsonPath":".status.cr_state"}"#,
+    printcolumn = r#"{ "name":"pool_status", "type":"string", "description":"Control plane pool status", "jsonPath":".status.pool_status"}"#,
+    printcolumn = r#"{ "name":"secure", "type":"boolean", "description":"encryption enabled", "jsonPath":".status.secure"}"#,
+    printcolumn = r#"{ "name":"capacity", "type":"integer", "format": "int64", "minimum" : "0", "description":"total bytes", "jsonPath":".status.capacity"}"#,
+    printcolumn = r#"{ "name":"used", "type":"integer", "format": "int64", "minimum" : "0", "description":"used bytes", "jsonPath":".status.used"}"#,
+    printcolumn = r#"{ "name":"available", "type":"integer", "format": "int64", "minimum" : "0", "description":"available bytes", "jsonPath":".status.available"}"#
+)]
+/// The pool spec which contains the parameters we use when creating the pool
+pub struct DiskPoolSpec {
+    /// The node the pool is placed on
+    pub node: String,
+    /// The disk device the pool is located on
+    pub disks: Vec<String>,
+    /// The topology for data placement.
+    pub topology: Option<Topology>,
+    /// Use to create encrypted pool.
+    #[serde(rename = "encryptionKeyConfig")]
+    pub encryption_key_config: Option<EncryptionKeyConfig>,
+}
+
+/// Placement pool topology used by volume operations.
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, JsonSchema)]
+pub struct Topology {
+    /// Label for topology
+    #[serde(default)]
+    pub labelled: HashMap<String, String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, JsonSchema)]
+pub struct EncryptionKeyConfig {
+    #[serde(rename = "type")]
+    pub source_type: EncryptionSourceType,
+    pub config: EncryptionConfig,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, JsonSchema)]
+pub enum EncryptionSourceType {
+    Secret,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, JsonSchema)]
+pub struct EncryptionConfig {
+    pub name: String,
```

Comment:
Shouldn't name be specific for secret?

Might suggest this instead?
```rust
    /// Use to create encrypted pool.
    #[serde(rename = "encryptionConfig")]
    pub encryption_config: Option<EncryptionConfig>,
...
#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, JsonSchema)]
#[serde(rename = 'camelCase')]
pub struct EncryptionConfig {
    pub source_type: EncryptionSourceType,
    pub secret_config: Option<EncryptionSecretConfig>,
}
```

Reply:
I tried to follow the OEP structure defined by @dsharma-dc. Also in this shouldn't  `pub secret_config: Option<EncryptionSecretConfig>`  be optional?

Reply:
I would suggest this.

```rust
/// Use to create encrypted pool.
#[serde(rename = "encryptionConfig")]
pub encryption_config: Option<EncryptionConfig>,
```

```rust
#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, JsonSchema)]
pub struct EncryptionConfig {
    pub source: EncryptionSource,
}

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, JsonSchema)]
pub enum EncryptionSource {
    Secret(EncryptionSecretConfig),
}

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, JsonSchema)]
pub struct EncryptionSecretConfig {
    pub name: String,
    pub namespace: String,
}
```

Reply:
Ideally yes this would be better, but I'm not sure if this type of rust enum is well supported for K8s definitions?

Reply:
Done. It worked as we tested, so kept it.